packages/trace-viewer/src/ui/llm.tsx

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import { EventEmitter } from '@testIsomorphic/events';
+import { useCookies } from '@web/uiUtils';
+
+export type LLMMessage = {
+  role: 'user' | 'assistant' | 'developer';
+  content: string;
+  displayContent?: string;
+};
+
+interface LLM {
+  readonly name: string;
+  chatCompletion(messages: LLMMessage[], signal: AbortSignal): AsyncGenerator<string>;
+}
+
+// https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream
+async function *parseSSE(body: NonNullable<Response['body']>): AsyncGenerator<{ type: string, data: string, id: string }> {
```

Comment:
I wonder why not use `EventSource`? Not that I don't like parsing things 😄

Reply:
Because EventSource doesn't support passing headers or POST requests :/